nistats/contrasts.py

```diff
@@ -0,0 +1,262 @@
+
+    dof_ = regression_result[label_].df_resid
+    return Contrast(effect=effect_, variance=var_, dof=dof_,
+                    contrast_type=contrast_type)
+
+
+def _fixed_effect_contrast(labels, results, con_vals, contrast_type=None):
+    """Computes the summary contrast assuming fixed effects.
+
+    Adds the same contrast applied to all labels and results lists.
+    """
+    contrast = None
+    for i, (lab, res, con_val) in enumerate(zip(labels, results, con_vals)):
+        if np.all(con_val == 0):
+            warn('Contrast for session %d is null' % i)
+            continue
+        contrast_ = compute_contrast(lab, res, con_val, contrast_type)
+        if contrast is None:
+            contrast = contrast_
+        else:
+            contrast = contrast + contrast_
+    if contrast is None:
+        raise ValueError('all contrasts provided were null contrasts')
+    return contrast
+
+
+class Contrast(object):
+    """ The contrast class handles the estimation of statistical contrasts
+    on a given model: student (t) or Fisher (F).
+    The important feature is that it supports addition,
+    thus opening the possibility of fixed-effects models.
+
+    The current implementation is meant to be simple,
+    and could be enhanced in the future on the computational side
+    (high-dimensional F constrasts may lead to memory breakage).
+    """
+
+    def __init__(self, effect, variance, dof=DEF_DOFMAX, contrast_type='t',
+                 tiny=DEF_TINY, dofmax=DEF_DOFMAX):
+        """
+        Parameters
+        ----------
+        effect : array of shape (contrast_dim, n_voxels)
+            the effects related to the contrast
+
+        variance : array of shape (contrast_dim, contrast_dim, n_voxels)
+            the associated variance estimate
+
+        dof : scalar
+            the degrees of freedom of the resiudals
+
+        contrast_type: {'t', 'F'}
+            specification of the contrast type
+        """
+        if variance.ndim != 3:
+            raise ValueError('Variance array should have 3 dimensions')
+        if effect.ndim != 2:
+            raise ValueError('Effect array should have 2 dimensions')
+        if variance.shape[0] != variance.shape[1]:
+            raise ValueError('Inconsistent shape for the variance estimate')
+        if ((variance.shape[1] != effect.shape[0]) or
+            (variance.shape[2] != effect.shape[1])):
+            raise ValueError('Effect and variance have inconsistent shape')
+
+        self.effect = effect
+        self.variance = variance
+        self.dof = float(dof)
+        self.dim = effect.shape[0]
+        if self.dim > 1 and contrast_type is 't':
+            print('Automatically converted multi-dimensional t to F contrast')
+            contrast_type = 'F'
+        self.contrast_type = contrast_type
+        self.stat_ = None
+        self.p_value_ = None
+        self.baseline = 0
+        self.tiny = tiny
+        self.dofmax = dofmax
+
+    def eff(self):
```

Comment:
then can them `effect()` and `variance()`.
If they are bound to be used only there, to avoid redundancy with the class parameters, I would have a unique `summary_statistics()` method that returns both.